hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/DefaultAddressProvider.java

```diff
@@ -18,37 +18,42 @@
 
 import com.hazelcast.client.config.ClientNetworkConfig;
 import com.hazelcast.client.connection.AddressProvider;
+import com.hazelcast.client.connection.Addresses;
 import com.hazelcast.client.util.AddressHelper;
 import com.hazelcast.nio.Address;
 
-import java.util.Collection;
-import java.util.LinkedList;
 import java.util.List;
 
 /**
  * Default address provider of Hazelcast.
- *
+ * <p>
  * Loads addresses from the Hazelcast configuration.
  */
 public class DefaultAddressProvider implements AddressProvider {
 
-    private ClientNetworkConfig networkConfig;
+    private final ClientNetworkConfig networkConfig;
 
     public DefaultAddressProvider(ClientNetworkConfig networkConfig) {
         this.networkConfig = networkConfig;
     }
 
     @Override
-    public Collection<Address> loadAddresses() {
-        final List<String> addresses = networkConfig.getAddresses();
-        if (addresses.isEmpty()) {
-            addresses.add("127.0.0.1");
+    public Addresses loadAddresses() {
+        List<String> configuredAddresses = networkConfig.getAddresses();
+
+        if (configuredAddresses.isEmpty()) {
+            configuredAddresses.add("127.0.0.1");
         }
-        final List<Address> possibleAddresses = new LinkedList<Address>();
 
-        for (String address : addresses) {
-            possibleAddresses.addAll(AddressHelper.getSocketAddresses(address));
+        Addresses addresses = new Addresses();
+        for (String address : configuredAddresses) {
+            List<Address> addressesForMember = AddressHelper.getSocketAddresses(address);
+            if (!addressesForMember.isEmpty()) {
+                addresses.primary().add(addressesForMember.remove(0));
```

Comment:
I think for simplicity you can even have `AddressHelper.getSocketAddresses` method return `Addresses` and do the primary secondary separation internally (It seemed to me that such generation can be the responsibility of `AddressHelper.getSocketAddresses`). I see that the method `AddressHelper.getSocketAddresses` is only called from this place which should be an easy change.